### PR TITLE
Optimize asset secret update logic.

### PR DESCRIPTION
**Pull request by mikechengwei**
Optimize asset secret update logic , when secret data equal to exist data , the asset can avoid updates and reduce the pressure on k8s apiserver.

```release-note:REPLACEME
Optimize asset secret update logic.
```

### CODE REVIEW

File: pkg/prometheus/operator.go [hunk 1]
```diff
@@ -1591,8 +1591,12 @@ func (c *Operator) createOrUpdateTLSAssetSecret(ctx context.Context, p *monitori
...
 
 	} else {
-		err = k8sutil.UpdateSecret(ctx, sClient, tlsAssetsSecret)
-		level.Debug(c.logger).Log("msg", "updated tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
+		eq := reflect.DeepEqual(secret.Data, tlsAssetsSecret.Data)
+		if !eq {
+			err = k8sutil.UpdateSecret(ctx, sClient, tlsAssetsSecret)
+			level.Debug(c.logger).Log("msg", "updated tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
+		}
+
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:1599]:
```suggestion
		if reflect.DeepEqual(secret.Data, tlsAssetsSecret.Data) {
			return nil
		}
		err = k8sutil.UpdateSecret(ctx, sClient, tlsAssetsSecret)
		level.Debug(c.logger).Log("msg", "updated tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
```

**Review by simonpasquier — COMMENTED** [review]
the same change should happen in pkg/alertmanager/operator.go (and maybe a good time to factorize some code).

**Comment by mikechengwei** [comment]
> the same change should happen in pkg/alertmanager/operator.go (and maybe a good time to factorize some code).
I have change code in `pkg/alertmanager/operator.go` again. Do you mean to factorize some code, for example, to extract the common code and put it in the util class?

For example , extract below code to a common method .
```
secret, err := sClient.Get(ctx, tlsAssetsSecret.Name, metav1.GetOptions{})
	if err != nil {
		if !apierrors.IsNotFound(err) {
			return errors.Wrapf(
				err,
				"failed to check whether tls assets secret already exists for Prometheus %v in namespace %v",
				p.Name,
				p.Namespace,
			)
		}
		_, err = sClient.Create(ctx, tlsAssetsSecret, metav1.CreateOptions{})
		level.Debug(c.logger).Log("msg", "created tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)

	} else {
		if reflect.DeepEqual(secret.Data, tlsAssetsSecret.Data) {
			return nil
		}
		err = k8sutil.UpdateSecret(ctx, sClient, tlsAssetsSecret)
		level.Debug(c.logger).Log("msg", "updated tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
	}

	if err != nil {
		return errors.Wrapf(err, "failed to create TLS assets secret for Prometheus %v in namespace %v", p.Name, p.Namespace)
	}
```

**Comment by simonpasquier** [comment]
> Do you mean to factorize some code, for example, to extract the common code and put it in the util class?

yes it would be neat :) I think that you can repurpose `k8sutil.UpdateSecret()` to be `k8sutil.CreateOrUpdateSecret()`: the latter can skip the update if existing and updated secrets are identical in terms of data and metadata (labels/annotations).

File: pkg/k8sutil/k8sutil.go [hunk 2]
```diff
@@ -175,20 +176,33 @@ func UpdateStatefulSet(ctx context.Context, sstClient clientappsv1.StatefulSetIn
...
+				desired.Name,
+				desired.Namespace,
+			)
+		}
+		_, err = secretClient.Create(ctx, desired, metav1.CreateOptions{})
 		return err
 	}
-
+	// If secret exist,
+	mutated := existingSecret.DeepCopyObject().(*v1.Secret)
```
Comment (simonpasquier) [inline pkg/k8sutil/k8sutil.go:195]: I don't think we need to deep-copy the object

File: pkg/k8sutil/k8sutil.go [hunk 3]
```diff
@@ -175,20 +176,33 @@ func UpdateStatefulSet(ctx context.Context, sstClient clientappsv1.StatefulSetIn
...
+
+	mergeMetadata(&desired.ObjectMeta, mutated.ObjectMeta)
+	mutated.Data = desired.Data
+	// diff and whether update
+	if !apiequality.Semantic.DeepEqual(existingSecret, mutated) {
```
Comment (simonpasquier) [inline pkg/k8sutil/k8sutil.go:200]: why not reflect.DeepEqual() on the Data fields?

File: pkg/k8sutil/k8sutil_test.go [hunk 4]
```diff
@@ -76,32 +76,6 @@ func TestMergeMetadata(t *testing.T) {
 		modifiedLabels      map[string]string
 		modifiedAnnotations map[string]string
 	}{
-		{
```
Comment (simonpasquier) [inline pkg/k8sutil/k8sutil_test.go:79]: why did you remove these test cases?

File: pkg/prometheus/operator.go [hunk 5]
```diff
@@ -1529,29 +1529,7 @@ func (c *Operator) createOrUpdateConfigurationSecret(ctx context.Context, p *mon
...
-			return nil
-		}
-		level.Debug(c.logger).Log("msg", "current Prometheus configuration has changed")
-	} else {
-		level.Debug(c.logger).Log("msg", "no current Prometheus configuration secret found", "currentConfigFound", curConfigFound)
-	}
-
-	level.Debug(c.logger).Log("msg", "updating Prometheus configuration secret")
-	return k8sutil.UpdateSecret(ctx, sClient, s)
+	return k8sutil.CreateOrUpdateSecret(ctx, sClient, s)
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:1532]:
```suggestion
        level.Debug(c.logger).Log("msg", "updating Prometheus configuration secret")
	return k8sutil.CreateOrUpdateSecret(ctx, sClient, s)
```

File: pkg/k8sutil/k8sutil.go (hunk 3 above)
Comment (mikechengwei) [reply pkg/k8sutil/k8sutil.go:199]: We also need to compare labels and annoations. `apiequality.Semantic.DeepEqual` method is like reflect.DeepEqual, but focused on semantic equality instead of memory equality.

File: pkg/k8sutil/k8sutil_test.go (hunk 4 above)
Comment (mikechengwei) [reply pkg/k8sutil/k8sutil_test.go]: Sorry , my test result.

File: pkg/k8sutil/k8sutil.go (hunk 2 above)
Comment (mikechengwei) [reply pkg/k8sutil/k8sutil.go]: We must use deepcopy a new secret from old secret , because `mergeMetadata` will change existingSecret value .

Comment (mikechengwei) [reply pkg/k8sutil/k8sutil.go:194]: We need to merge a new secret to compare with old secret.

File: pkg/k8sutil/k8sutil.go (hunk 3 above)
Comment (simonpasquier) [reply pkg/k8sutil/k8sutil.go]: :+1: looking quickly at the definition of `apiequality.Semantic.DeepEqual()`, I thought that it was only checking a few fields but now I see that it uses reflect.DeepEqual() when needed underneath :)

File: pkg/k8sutil/k8sutil.go [hunk 6]
```diff
@@ -175,20 +176,30 @@ func UpdateStatefulSet(ctx context.Context, sstClient clientappsv1.StatefulSetIn
...
-
+	mutated := existingSecret.DeepCopyObject().(*v1.Secret)
+	mergeMetadata(&desired.ObjectMeta, mutated.ObjectMeta)
+	// diff and whether update
+	if !apiequality.Semantic.DeepEqual(existingSecret, desired) {
+		_, err = secretClient.Update(ctx, desired, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
```
Comment (simonpasquier) [inline pkg/k8sutil/k8sutil.go:202]: Lets add secret name/namespace. Also a bit nicer to read IMHO (less nested levels).

```suggestion
	if apiequality.Semantic.DeepEqual(existingSecret, desired) {
                return nil
        }
	
	if _, err = secretClient.Update(ctx, desired, metav1.UpdateOptions{}); err != nil {
		return errors.Wrapf(
			err,
			"failed to update secret %q in namespace %q",
			desired.Name,
			desired.Namespace,
		)
	}
	
	return nil
```

File: pkg/k8sutil/k8sutil.go [hunk 7]
```diff
@@ -175,20 +176,30 @@ func UpdateStatefulSet(ctx context.Context, sstClient clientappsv1.StatefulSetIn
...
+		if !apierrors.IsNotFound(err) {
+			return errors.Wrapf(
+				err,
+				"failed to check whether secret %q in namespace %q already exists",
+				desired.Name,
+				desired.Namespace,
+			)
+		}
+		_, err = secretClient.Create(ctx, desired, metav1.CreateOptions{})
 		return err
```
Comment (simonpasquier) [inline pkg/k8sutil/k8sutil.go:192]: lets wrap here too

```suggestion
		return errors.Wrapf(err, "failed to create secret %q in namespace %q", desired.Name, desired.Namespace)
```

File: pkg/alertmanager/operator.go [hunk 8]
```diff
@@ -830,23 +830,7 @@ func (c *Operator) createOrUpdateGeneratedConfigSecret(ctx context.Context, am *
...
-		_, err = sClient.Create(ctx, generatedConfigSecret, metav1.CreateOptions{})
-		level.Debug(c.logger).Log("msg", "created generated config secret", "secretname", generatedConfigSecret.Name)
-	} else {
-		err = k8sutil.UpdateSecret(ctx, sClient, generatedConfigSecret)
-		level.Debug(c.logger).Log("msg", "updated generated config secret", "secretname", generatedConfigSecret.Name)
-	}
-
+	err := k8sutil.CreateOrUpdateSecret(ctx, sClient, generatedConfigSecret)
 	if err != nil {
 		return errors.Wrapf(err, "failed to update generated config secret for Alertmanager %v in namespace %v", am.Name, am.Namespace)
```
Comment (simonpasquier) [inline pkg/alertmanager/operator.go:835]: after looking at the code I think we can remove the name and namespace of the Alertmanager resource because it s already added by the caller (e.g. `sync()`):

```suggestion
		return errors.Wrap(err, "failed to update generated config secret")
```

File: pkg/alertmanager/operator.go [hunk 9]
```diff
@@ -1313,24 +1297,7 @@ func (c *Operator) createOrUpdateTLSAssetSecret(ctx context.Context, am *monitor
...
-		level.Debug(c.logger).Log("msg", "created tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
-
-	} else {
-		err = k8sutil.UpdateSecret(ctx, sClient, tlsAssetsSecret)
-		level.Debug(c.logger).Log("msg", "updated tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
-	}
-
+	err := k8sutil.CreateOrUpdateSecret(ctx, sClient, tlsAssetsSecret)
 	if err != nil {
 		return errors.Wrapf(err, "failed to create TLS assets secret for Alertmanager %v in namespace %v", am.Name, am.Namespace)
```
Comment (simonpasquier) [inline pkg/alertmanager/operator.go:1302]: ditto

File: pkg/prometheus/operator.go [hunk 10]
```diff
@@ -1580,24 +1559,7 @@ func (c *Operator) createOrUpdateTLSAssetSecret(ctx context.Context, p *monitori
...
-		level.Debug(c.logger).Log("msg", "created tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
-
-	} else {
-		err = k8sutil.UpdateSecret(ctx, sClient, tlsAssetsSecret)
-		level.Debug(c.logger).Log("msg", "updated tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
-	}
-
+	err := k8sutil.CreateOrUpdateSecret(ctx, sClient, tlsAssetsSecret)
 	if err != nil {
 		return errors.Wrapf(err, "failed to create TLS assets secret for Prometheus %v in namespace %v", p.Name, p.Namespace)
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:1564]: ditto about removing name and namespace

File: pkg/prometheus/operator.go [hunk 11]
```diff
@@ -1580,24 +1559,7 @@ func (c *Operator) createOrUpdateTLSAssetSecret(ctx context.Context, p *monitori
...
-		}
-		_, err = sClient.Create(ctx, tlsAssetsSecret, metav1.CreateOptions{})
-		level.Debug(c.logger).Log("msg", "created tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
-
-	} else {
-		err = k8sutil.UpdateSecret(ctx, sClient, tlsAssetsSecret)
-		level.Debug(c.logger).Log("msg", "updated tlsAssetsSecret", "secretname", tlsAssetsSecret.Name)
-	}
-
+	err := k8sutil.CreateOrUpdateSecret(ctx, sClient, tlsAssetsSecret)
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:1562]: we can merge with the next block

```suggestion
	if err := k8sutil.CreateOrUpdateSecret(ctx, sClient, tlsAssetsSecret); err != nil {
```

**Review by simonpasquier — COMMENTED** [review]
Thanks for your patience :) just a few minor comments to make the code more consistent but overall :+1: for me.

**Comment by mikechengwei** [comment]
@simonpasquier Thank you very much for your guidance and learn together. Please help me  review again.

**Comment by mikechengwei** [comment]
@simonpasquier

**Review by simonpasquier — APPROVED** [review]

**Comment by simonpasquier** [comment]
Looks good to me, thanks a lot!
